### PR TITLE
GTEST: Add test analysis at the end of testing

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

1. Add the infrastructure to report the TOP-N longest tests from the gtest environment
2. Use this infrastructure from Jenkins to report TOP-20 longest tests

## Why ?

This is needed to always be aware of the longest tests and use for the development purposes

## How ?

Collect the name from `UnitTest::TestCase::TestInfo` and the testing time from `UnitTest::TestCase::TestInfo::TestResut`, store them in the vector, sort the vector na dreport TOP-N longest tests.

### CODE REVIEW

**Comment by dmitrygladkov** [comment]
This is an example from bgate Jenkins testing:

```
14:25:30 TOP-20 longest tests:
14:25:30 1. cma/uct_p2p_rma_test.get_zcopy/0                                                    - 65568 ms
14:25:30 2. mm/uct_p2p_rma_test.put_bcopy/0                                                     - 48827 ms
14:25:30 3. mm/uct_p2p_rma_test.get_bcopy/2                                                     - 44882 ms
14:25:30 4. mm/uct_p2p_rma_test.put_bcopy/4                                                     - 44302 ms
14:25:30 5. knem/uct_p2p_rma_test.get_zcopy/0                                                   - 43632 ms
14:25:30 6. mm/uct_p2p_rma_test.put_short/2                                                     - 40126 ms
14:25:30 7. tcp/uct_p2p_am_misc.no_rx_buffs/0                                                   - 39230 ms
14:25:30 8. tcp/test_uct_pending.pending_op/0                                                   - 34518 ms
14:25:30 9. tcp/test_ucp_tag_perf.multi_unexp/0                                                 - 30462 ms
14:25:30 10. rc/uct_p2p_rma_test_alloc_methods.xfer_reg_odp/1                                    - 25321 ms
14:25:30 11. tcp/test_ucp_tag_xfer.test_xfer_len_offset/3                                        - 25041 ms
14:25:30 12. rcx/test_ucp_perf.envelope/0                                                        - 23961 ms
14:25:30 13. dc_mlx5/test_uct_peer_failure_multiple.test/0                                       - 20302 ms
14:25:30 14. rc_mlx5/uct_p2p_rma_test_alloc_methods.xfer_reg_rcache/1                            - 18741 ms
14:25:30 15. dc_mlx5/uct_p2p_rma_test_alloc_methods.xfer_reg_rcache/1                            - 18368 ms
14:25:30 16. rc/uct_p2p_rma_test_alloc_methods.xfer_reg_direct/1                                 - 18270 ms
14:25:30 17. tcp/test_ucp_tag_match.rndv_exp_huge_mix/0                                          - 18142 ms
14:25:30 18. signal/test_async.many_timers/0                                                     - 10714 ms
14:25:30 19. udx/test_ucp_tag_match.rndv_exp_huge_mix/0                                          - 9934 ms
14:25:30 20. dc_mlx5/test_uct_peer_failure.two_pairs_send_after/0                                - 9892 ms
```

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10291/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10294/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
> Test FAILed.
> See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10294/ for details (Mellanox internal link).

related to #3607

**Comment by dmitrygladkov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10297/ for details (Mellanox internal link).

File: test/gtest/common/test_helpers.h [hunk 1]
```diff
@@ -170,7 +171,7 @@
 
 
 namespace ucs {
-
+
```
Comment (brminich) [inline test/gtest/common/test_helpers.h]: extra spaces?

Comment (dmitrygladkov) [reply test/gtest/common/test_helpers.h]: yes, fixed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10310/ for details (Mellanox internal link).

File: test/gtest/common/test_helpers.h [hunk 2]
```diff
@@ -193,6 +194,8 @@ typedef struct {
     int                   kill_signal;
 } test_watchdog_t;
 
+typedef std::pair<std::string, ::testing::TimeInMillis> test_result_t;
```
Comment (yosefe) [inline test/gtest/common/test_helpers.h]: can we define it in CC file only? it's private helper type

File: contrib/test_jenkins.sh [hunk 3]
```diff
@@ -1047,8 +1047,10 @@ run_gtest() {
 	export GTEST_SHUFFLE=1
 	export GTEST_TAP=2
 	export GTEST_REPORT_DIR=$WORKSPACE/reports/tap
-	# Run UCT tests for TCP over RDMA devices only
+	# Run UCT tests for TCP over fastest device only
 	export GTEST_UCT_TCP_FASTEST_DEV=1
+	# Report TOP-20 longest test at the end of testing
+	export GTEST_ANALYZE_RESULTS=20
```
Comment (yosefe) [inline contrib/test_jenkins.sh]: GTEST_REPORT_LONGEST_TESTS

File: test/gtest/common/test_helpers.cc [hunk 4]
```diff
@@ -208,6 +208,88 @@ void watchdog_stop()
...
+void analyze_test_results()
+{
+    // GTEST_ANALYZE_RESULTS=100 will report TOP-100 longest tests
+    /* coverity[tainted_data_return] */
+    char *env_p = getenv("GTEST_ANALYZE_RESULTS");
+    if (env_p == NULL) {
+        return;
+    }
+
+    int max_name_size = 0, top_n;
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc]: make max_name_size size_t

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10314/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10316/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@yosefe @brminich do you have more comments?

File: test/gtest/common/test_helpers.cc [hunk 5]
```diff
@@ -208,6 +210,89 @@ void watchdog_stop()
...
+
+    std::sort(test_results.begin(), test_results.end(), test_results_cmp);
+
+    top_n = std::min((int)test_results.size(), top_n);
+    if (!top_n) {
+        return;
+    }
+
+    int max_index_size = ucs::to_string(top_n).size();
+    std::cout << std::endl << "TOP-" << top_n << " longest tests:" << std::endl;
```
Comment (brminich) [inline test/gtest/common/test_helpers.cc]: Will this string be printed only, if negative digit is provided for top_n?

Comment (dmitrygladkov) [reply test/gtest/common/test_helpers.cc]: yes, you're right
do we need to add sanity checks for the value passed to `GTEST_REPORT_LONGEST_TESTS`?

Comment (yosefe) [reply test/gtest/common/test_helpers.cc]: i don't think so, it's internal/test stuff

Comment (brminich) [reply test/gtest/common/test_helpers.cc]: I'd just ignore negative values (check for >0 instead of !=0), but it's too late anyway

**Pull request merged**